src/iter.rs

```diff
@@ -1125,43 +767,468 @@ impl<'a> Chunks<'a> {
+                let (node, ref mut child_i) = node_stack.last_mut().unwrap();
+
+                // Get the text, sliced to the appropriate range.
+                let text = node.children().nodes()[*child_i].leaf_text();
+                let text_slice = {
+                    let start_byte = if *byte_idx < 0 {
+                        (-*byte_idx) as usize
+                    } else {
+                        0
+                    };
+                    let end_byte = text.len().min((total_bytes as isize - *byte_idx) as usize);
+                    &text[start_byte..end_byte]
+                };
+
+                // Book keeping.
+                *byte_idx += text.len() as isize;
+                *child_i += 1;
+
+                // Return the text.
+                return Some(text_slice);
+            }
+
+            Chunks {
+                iter:
+                    ChunksEnum::Light {
+                        text,
+                        ref mut is_end,
+                    },
+                ..
+            } => {
+                if *is_end || text.is_empty() {
+                    return None;
+                } else {
+                    *is_end = true;
+                    return Some(text);
+                }
+            }
+        }
+    }
+}
+
+impl<'a> Iterator for Chunks<'a> {
+    type Item = &'a str;
+
+    /// Advances the iterator forward and returns the next value.
+    ///
+    /// Runs in amortized O(1) time and worst-case O(log N) time.
+    #[inline(always)]
+    fn next(&mut self) -> Option<&'a str> {
+        if !self.is_reversed {
+            self.next_impl()
+        } else {
+            self.prev_impl()
+        }
+    }
+}
+
+//==========================================================
+
+/// An iterator over a `Rope`'s contiguous `str` chunks.
+///
+/// Internally, each `Rope` stores text as a segemented collection of utf8
+/// strings. This iterator iterates over those segments, returning a
+/// `&str` slice for each one.  It is useful for situations such as:
+///
+/// - Writing a rope's utf8 text data to disk (but see
+///   [`write_to()`](crate::rope::Rope::write_to) for a convenience function that does this
+///   for casual use-cases).
+/// - Streaming a rope's text data somewhere.
+/// - Saving a rope to a non-utf8 encoding, doing the encoding conversion
+///   incrementally as you go.
+/// - Writing custom iterators over a rope's text data.
+///
+/// There are precisely two guarantees about the yielded chunks:
+///
+/// - All non-empty chunks are yielded, and they are yielded in order.
+/// - CRLF pairs are never split across chunks.
+///
+/// There are no guarantees about the size of yielded chunks, and except for
+/// CRLF pairs and being valid `str` slices there are no guarantees about
+/// where the chunks are split.  For example, they may be zero-sized, they
+/// don't necessarily align with line breaks, etc.
+#[derive(Debug, Clone)]
+pub struct Lines<'a> {
```

Comment:
It looks like you accidentally copy-pasted the `Chunks` iterator documentation, overwritten the correct `Lines` iterator documentation.

Reply:
Yes you are right, thats the case for all functions. I started off with the chunks iterator as a blueprint andd totally forgot about fixing that! I am not home right now but I will fix this first thing tomorrow morning

Reply:
It also looks like you moved the `Lines` iterator to be below the `Chunks` iterator, which seems to be adding to the noisiness of the diff on Github.  If you could move it back (the order should be: `Bytes`, `Chars`, `Lines`, `Chunks`) I'd really appreciate it.  I don't really care about the order per se, but it would make this already-challenging PR a little easier to review.

(And just in general, trying to minimize irrelevant changes in PRs is usually appreciated by reviewers, regardless of project.)

Reply:
> I am not home right now but I will fix this first thing tomorrow morning

No rush!  This PR is going to take a while for me to fully review, because of how involved it is.

Reply:
I have fixed the documentation (and also updated it to reflect the now improved time complexity) and moved the `Lines` iterator back to the previous position. A git diff will not be all that useful tough for this PR because ther is basically no carry-over from the old implementation (the `Lines` iterator is closer now to the `Chunks` iterator then the old `Lines` iterator) so I woud suggest just looking at the implementation instead of the diff while reviewing

Reply:
Yeah, I don't expect the diff to be *especially* useful.  But when scrolling through to make comments, I was having to scroll all the way past the "added" Chunks iterator (which was actually just moved).  I also don't expect the "light" variant of the code to really be meaningfully different, and this makes that easier to look at as well.